packages/flutter_tools/lib/src/mdns_discovery.dart

```diff
@@ -146,20 +148,69 @@ class MDnsObservatoryDiscovery {
       applicationId: applicationId,
       deviceVmservicePort: deviceVmservicePort,
     );
-    Uri observatoryUri;
-    if (result != null) {
-      final String host = usesIpv6
-        ? InternetAddress.loopbackIPv6.address
-        : InternetAddress.loopbackIPv4.address;
-      observatoryUri = await buildObservatoryUri(
-        device,
-        host,
-        result.port,
-        hostVmservicePort,
-        result.authCode,
-      );
+    if (result == null) {
+      await _checkForIPv4LinkLocal(device);
+      return null;
+    }
+
+    final String host = usesIpv6
+      ? InternetAddress.loopbackIPv6.address
+      : InternetAddress.loopbackIPv4.address;
+    return await buildObservatoryUri(
+      device,
+      host,
+      result.port,
+      hostVmservicePort,
+      result.authCode,
+    );
+  }
+
+  // If there's not an ipv4 link local address in `NetworkInterfaces.list`,
+  // then request user interventions with a `printError()` if possible.
+  Future<void> _checkForIPv4LinkLocal(Device device) async {
+    printTrace(
+      'mDNS query failed. Checking for an interface with a ipv4 link local address.'
+    );
+    final List<NetworkInterface> interfaces = await listNetworkInterfaces(
+      includeLinkLocal: true,
+      includeLoopback: true,
+      type: InternetAddressType.IPv4,
+    );
+    if (logger.isVerbose) {
+      _logInterfaces(interfaces);
+    }
+    final bool hasIPv4LinkLocal = interfaces.any(
+      (NetworkInterface interface) => interface.addresses[0].isLinkLocal);
```

Comment:
This probably works, but maybe `interface.addresses.any((InternetAddress address) => address.isLinkLocal)`?

Reply:
I don't know, I was mainly cribbing from `package:multicast_dns`. It binds the sending socket to the first address on each interface. If that address is not a link local address on an interface, could that be the reason that an mDNS query fails?

Reply:
Went with this change.

Reply:
Yeah, I think this should normally be fine, and it's not handled "properly" in multicast_dns